File: Vb/ImageProcessing/ImgProc/Binarization.vb
Imports System.Threading
Imports System.Runtime.InteropServices.Marshal
Imports System.Drawing.Imaging

Public Class Binarization : Inherits ComImgProc
    Private m_nThresh As Byte

    Public Property Thresh() As Byte
        Set(value As Byte)
            m_nThresh = value
        End Set
        Get
            Return m_nThresh
        End Get
    End Property

    Public Sub New(_bitmap As Bitmap)
        MyBase.New(_bitmap)

        m_nThresh = 0
    End Sub

    Public Sub New(_bitmap As Bitmap, _nThresh As Byte)
        MyBase.New(_bitmap)

        m_nThresh = _nThresh
    End Sub

    Protected Overrides Sub Finalize()
        MyBase.Finalize()
    End Sub

    Public Overrides Function GoImgProc(_token As CancellationToken) As Boolean
        Dim bRst As Boolean = True

        Dim nWidthSize As Integer = MyBase.m_bitmap.Width
        Dim nHeightSize As Integer = MyBase.m_bitmap.Height

        Dim bitmapData As BitmapData = MyBase.m_bitmap.LockBits(New Rectangle(0, 0, nWidthSize, nHeightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

        Dim nIdxWidth As Integer
        Dim nIdxHeight As Integer

        For nIdxHeight = 0 To nHeightSize - 1 Step 1
            For nIdxWidth = 0 To nWidthSize - 1 Step 1
                If (_token.IsCancellationRequested = True) Then
                    bRst = False
                    Exit For
                End If

                Dim pAdr As IntPtr = bitmapData.Scan0
                Dim nPos As Integer = nIdxHeight * bitmapData.Stride + nIdxWidth * 4
                Dim nPixelB As Integer = ReadByte(pAdr, nPos + ComInfo.Pixel.B)
                Dim nPixelG As Integer = ReadByte(pAdr, nPos + ComInfo.Pixel.G)
                Dim nPixelR As Integer = ReadByte(pAdr, nPos + ComInfo.Pixel.R)

                Dim nGrayScale As Integer = (nPixelB + nPixelG + nPixelR) / 3

                Dim nBinarization As Byte = If(nGrayScale >= m_nThresh, CByte(255), CByte(0))
                WriteByte(pAdr, nPos + ComInfo.Pixel.B, CByte(nBinarization))
                WriteByte(pAdr, nPos + ComInfo.Pixel.G, CByte(nBinarization))
                WriteByte(pAdr, nPos + ComInfo.Pixel.R, CByte(nBinarization))
            Next
        Next
        MyBase.m_bitmap.UnlockBits(bitmapData)

        Return bRst
    End Function
End Class

File: CSharp/ImageProcessing/ImgProc/Binarization.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;

/// <summary>
/// 2値化のロジック
/// </summary>
class Binarization : ComImgProc
{
    private byte m_nThresh;

    /// <summary>
    /// 閾値
    /// </summary>
    public byte Thresh
    {
        set { m_nThresh = value; }
        get { return m_nThresh; }
    }

    /// <summary>
    /// コンストラクタ
    /// </summary>
    /// <param name="_bitmap">ビットマップ</param>
    public Binarization(Bitmap _bitmap) : base(_bitmap)
    {
        m_nThresh = 0;
    }

    /// <summary>
    /// コンストラクタ
    /// </summary>
    /// <param name="_bitmap">ビットマップ</param>
    /// <param name="_nThresh">閾値</param>
    public Binarization(Bitmap _bitmap, byte _nThresh) : base(_bitmap)
    {
        m_nThresh = _nThresh;
    }

    /// <summary>
    /// 初期化
    /// </summary>
    public override void Init()
    {
        m_nThresh = 0;
        base.Init();
    }

    /// <summary>
    /// 2値化の実行
    /// </summary>
    /// <param name="_token">キャンセルトークン</param>
    /// <returns>実行結果 成功/失敗</returns>
    public override bool GoImgProc(CancellationToken _token)
    {
        bool bRst = true;

        int nWidthSize = base.m_bitmap.Width;
        int nHeightSize = base.m_bitmap.Height;

        base.m_bitmapAfter = new Bitmap(base.m_bitmap);

        BitmapData bitmapData = base.m_bitmapAfter.LockBits(new Rectangle(0, 0, nWidthSize, nHeightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);

        int nIdxWidth;
        int nIdxHeight;

        unsafe
        {
            for (nIdxHeight = 0; nIdxHeight < nHeightSize; nIdxHeight++)
            {
                if (_token.IsCancellationRequested)
                {
                    bRst = false;
                    break;
                }

                for (nIdxWidth = 0; nIdxWidth < nWidthSize; nIdxWidth++)
                {
                    if (_token.IsCancellationRequested)
                    {
                        bRst = false;
                        break;
                    }

                    byte* pPixel = (byte*)bitmapData.Scan0 + nIdxHeight * bitmapData.Stride + nIdxWidth * 4;
                    byte nGrayScale = (byte)((pPixel[(int)ComInfo.Pixel.B] + pPixel[(int)ComInfo.Pixel.G] + pPixel[(int)ComInfo.Pixel.R]) / 3);

                    byte nBinarization = nGrayScale >= m_nThresh ? (byte)255 : (byte)0;
                    pPixel[(int)ComInfo.Pixel.B] = nBinarization;
                    pPixel[(int)ComInfo.Pixel.G] = nBinarization;
                    pPixel[(int)ComInfo.Pixel.R] = nBinarization;
                }
            }
            base.m_bitmapAfter.UnlockBits(bitmapData);
        }

        return bRst;
    }
}